### PR TITLE
Feature 5163 configurable socket keepalive interval

### PR DESCRIPTION
**Pull request by aksabg**
Feature according to: [github.com/eclipse-vertx/vert.x/issues/5163](github.com/eclipse-vertx/vert.x/issues/5163)

### CODE REVIEW

File: src/main/java/io/vertx/core/net/NetServerOptions.java [hunk 1]
```diff
@@ -531,6 +554,62 @@ public NetServerOptions setTrafficShapingOptions(TrafficShapingOptions trafficSh
     return this;
   }
 
+  /**
+   * @return the time in seconds the connection needs to remain idle before TCP starts sending keepalive probes
+   */
+  public int getTcpKeepAliveIdleSeconds() {
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/NetServerOptions.java:560]: why not make this part of TCPSSLOptions instead ? it does not seem specific to TCP server

File: src/main/java/io/vertx/core/spi/transport/Transport.java [hunk 2]
```diff
@@ -174,6 +174,15 @@ default void configure(NetServerOptions options, boolean domainSocket, ServerBoo
     bootstrap.option(ChannelOption.SO_REUSEADDR, options.isReuseAddress());
     if (!domainSocket) {
       bootstrap.childOption(ChannelOption.SO_KEEPALIVE, options.isTcpKeepAlive());
+      if (options.isTcpKeepAlive() && options.getTcpKeepAliveIdleSeconds() != -1) {
+        bootstrap.childOption(EpollChannelOption.TCP_KEEPIDLE, options.getTcpKeepAliveIdleSeconds());
```
Comment (vietj) [inline src/main/java/io/vertx/core/spi/transport/Transport.java:178]: Epoll options must be configured in EpollTransport subclass and not in this one, further more that forces users that do not use epoll to have the native epoll jar on their classpath

Comment (aksabg) [reply src/main/java/io/vertx/core/spi/transport/Transport.java:178]: Done

File: src/main/java/io/vertx/core/net/NetServerOptions.java (hunk 1 above)
Comment (aksabg) [reply src/main/java/io/vertx/core/net/NetServerOptions.java:560]: Done

File: src/main/java/io/vertx/core/impl/transports/EpollTransport.java [hunk 3]
```diff
@@ -145,6 +145,16 @@ public void configure(NetServerOptions options, boolean domainSocket, ServerBoot
       }
       bootstrap.childOption(EpollChannelOption.TCP_QUICKACK, options.isTcpQuickAck());
       bootstrap.childOption(EpollChannelOption.TCP_CORK, options.isTcpCork());
+
+      if (options.isTcpKeepAlive() && options.getTcpKeepAliveIdleSeconds() != -1) {
+        bootstrap.childOption(EpollChannelOption.TCP_KEEPIDLE, options.getTcpKeepAliveIdleSeconds());
+      }
+      if (options.isTcpKeepAlive() && options.getTcpKeepAliveCount() != -1) {
+        bootstrap.childOption(EpollChannelOption.TCP_KEEPCNT, options.getTcpKeepAliveIdleSeconds());
```
Comment (jakub-id) [inline src/main/java/io/vertx/core/impl/transports/EpollTransport.java:153]: Should be `getTcpKeepAliveCount`

File: src/main/java/io/vertx/core/impl/transports/EpollTransport.java [hunk 4]
```diff
@@ -145,6 +145,16 @@ public void configure(NetServerOptions options, boolean domainSocket, ServerBoot
...
+      }
+      if (options.isTcpKeepAlive() && options.getTcpKeepAliveIntervalSeconds() != -1) {
+        bootstrap.childOption(EpollChannelOption.TCP_KEEPINTVL, options.getTcpKeepAliveIdleSeconds());
```
Comment (jakub-id) [inline src/main/java/io/vertx/core/impl/transports/EpollTransport.java:156]: Should be `getTcpKeepAliveIntervalSeconds`

**Review by jakub-id — CHANGES_REQUESTED** [review]

File: src/main/java/io/vertx/core/impl/transports/EpollTransport.java (hunk 3 above)
Comment (aksabg) [reply src/main/java/io/vertx/core/impl/transports/EpollTransport.java:153]: Fixed

**Comment by vietj** [comment]
@aksabg can you signe the eclipse agreement ?

also can you avoid reformating code where it is not needed ?

**Comment by aksabg** [comment]
@vietj Done

**Comment by vietj** [comment]
@aksabg thanks for contributing this

**Pull request merged**